Sources/PackageModelSyntax/SyntaxEditUtils.swift

```diff
@@ -0,0 +1,225 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2014-2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Basics
+import PackageModel
+import SwiftSyntax
+import SwiftParser
+
+/// Default indent when we have to introduce indentation but have no context
+/// to get it right.
+let defaultIndent = TriviaPiece.spaces(4)
+
+extension Trivia {
+    /// Determine whether this trivia has newlines or not.
+    var hasNewlines: Bool {
+        contains { piece in
+            if case .newlines = piece {
+                return true
+            } else {
+                return false
+            }
+        }
+    }
+}
+
+/// Syntax walker to find the first occurrence of a given node kind that
+/// matches a specific predicate.
+private class FirstNodeFinder<Node: SyntaxProtocol>: SyntaxAnyVisitor {
+    var predicate: (Node) -> Bool
+    var found: Node? = nil
+
+    init(predicate: @escaping (Node) -> Bool) {
+        self.predicate = predicate
+        super.init(viewMode: .sourceAccurate)
+    }
+
+    override func visitAny(_ node: Syntax) -> SyntaxVisitorContinueKind {
+        if found != nil {
+            return .skipChildren
+        }
+
+        if let matchedNode = node.as(Node.self), predicate(matchedNode) {
+            found = matchedNode
+            return .skipChildren
+        }
+
+        return .visitChildren
+    }
+}
+
+extension SyntaxProtocol {
+    /// Find the first node of the Self type that matches the given predicate.
+    static func findFirst(
+        in node: some SyntaxProtocol,
+        matching predicate: (Self) -> Bool
+    ) -> Self? {
```

Comment:
Wouldn’t it make more sense for this to be an instance function on `SyntaxProtocol` instead of a static function?

Reply:
No, because you are looking for something of the `Self` type.

Reply:
I was thinking if it is a cleaner API to have it be like

```swift
tree.findFirst { (node: FunctionCallSyntax) -> Bool in
  doChecks(for: node)
}
```

Instead of 
```swift
FunctionCallSyntax.findFirst(in: tree) { node in
  doChecks(for: node)
}
```
which reads a little backwards to me. But maybe it’s a matter of taste.

Reply:
I hate requiring type annotations on closures... but I can create an `ofType` argument or similar to make this cleaner.